File: services/swift/order-service/Sources/App/Controllers/OrderController.swift
import Vapor
import Fluent

struct OrderController: RouteCollection {
    let productServiceURL: String
    let userServiceURL: String

    func boot(routes: any RoutesBuilder) throws {
        routes.get(use: list)
        routes.post(use: create)
        routes.group(":orderID") { r in
            r.get(use: get)
            r.patch("status", use: updateStatus)
        }
    }

    func list(req: Request) async throws -> [Order] {
        try await Order.query(on: req.db).with(\.$items).sort(\.$createdAt, .descending).limit(50).all()
    }

    func create(req: Request) async throws -> Response {
        let body = try req.content.decode(CreateOrderRequest.self)

        // Fetch product prices from product-service
        var total: Double = 0
        var items: [(UUID, Int, Double)] = []
        for item in body.items {
            let url = "\(productServiceURL)/products/\(item.productId)"
            let res = try await req.client.get(URI(string: url))
            guard res.status == .ok,
                  let price = try? res.content.get(Double.self, at: "price") else {
                throw Abort(.badGateway, reason: "Could not fetch product \(item.productId)")
            }
            let lineTotal = price * Double(item.quantity)
            total += lineTotal
            items.append((item.productId, item.quantity, price))
        }

        let order = Order(userId: body.userId, status: .pending, total: total)
        try await order.save(on: req.db)

        for (pid, qty, price) in items {
            let oi = OrderItem(productId: pid, quantity: qty, unitPrice: price)
            oi.$order.id = order.id!
            try await oi.save(on: req.db)
        }

        try await order.$items.load(on: req.db)
        return try await order.encodeResponse(status: .created, for: req)
    }

    func get(req: Request) async throws -> Order {
        guard let id = req.parameters.get("orderID", as: UUID.self) else { throw Abort(.badRequest) }
        guard let order = try await Order.query(on: req.db).filter(\.$id == id).with(\.$items).first() else {
            throw Abort(.notFound)
        }
        return order
    }

    func updateStatus(req: Request) async throws -> Order {
        guard let id = req.parameters.get("orderID", as: UUID.self) else { throw Abort(.badRequest) }
        guard let order = try await Order.find(id, on: req.db) else { throw Abort(.notFound) }
        let newStatus = try req.content.get(OrderStatus.self, at: "status")
        order.status = newStatus
        try await order.save(on: req.db)
        return order
    }
}

File: services/java/order-service/src/main/java/com/research/binus/orderservice/controller/OrderController.java
package com.research.binus.orderservice.controller;

import com.research.binus.orderservice.model.*;
import com.research.binus.orderservice.repository.OrderRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.server.ResponseStatusException;

import java.util.*;

@RestController
@RequiredArgsConstructor
public class OrderController {

    private final OrderRepository orderRepository;
    private final RestTemplate restTemplate;

    @Value("${services.product-url}")
    private String productServiceUrl;

    @GetMapping("/health")
    public Map<String, String> health() {
        return Map.of("status", "ok", "service", "java-order-service", "language", "java");
    }

    @GetMapping("/orders")
    public List<Order> list() {
        return orderRepository.findTop50ByOrderByCreatedAtDesc();
    }

    @PostMapping("/orders")
    public ResponseEntity<Order> create(@RequestBody CreateOrderRequest req) {
        double total = 0;
        List<OrderItem> items = new ArrayList<>();

        for (CreateOrderRequest.ItemRequest ir : req.items()) {
            String url = productServiceUrl + "/products/" + ir.productId();
            try {
                @SuppressWarnings("unchecked")
                Map<String, Object> product = restTemplate.getForObject(url, Map.class);
                double price = ((Number) product.get("price")).doubleValue();
                total += price * ir.quantity();
                items.add(OrderItem.builder()
                        .productId(ir.productId())
                        .quantity(ir.quantity())
                        .unitPrice(price)
                        .build());
            } catch (Exception e) {
                throw new ResponseStatusException(HttpStatus.BAD_GATEWAY, "Cannot reach product service");
            }
        }

        Order order = Order.builder()
                .userId(req.userId())
                .total(total)
                .status(OrderStatus.PENDING)
                .build();

        Order saved = orderRepository.save(order);
        for (OrderItem item : items) {
            item.setOrder(saved);
        }
        saved.getItems().addAll(items);
        Order finalOrder = orderRepository.save(saved);

        return ResponseEntity.status(HttpStatus.CREATED).body(finalOrder);
    }

    @GetMapping("/orders/{id}")
    public Order get(@PathVariable UUID id) {
        return orderRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));
    }

    @PatchMapping("/orders/{id}/status")
    public Order updateStatus(@PathVariable UUID id, @RequestBody Map<String, String> req) {
        Order order = orderRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));
        order.setStatus(OrderStatus.valueOf(req.get("status").toUpperCase()));
        return orderRepository.save(order);
    }

    record CreateOrderRequest(UUID userId, List<ItemRequest> items) {
        record ItemRequest(UUID productId, int quantity) {}
    }
}
